### PR TITLE
WIP RFC-15 Dynamic attribute and command

### PR DESCRIPTION
**Pull request by HelgaMond**
This RFC covers dynamic attribute and command features in Tango V9.

### CODE REVIEW

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: I find "server client code" a bit confusing here even though it is explained in the diagram below and in the definitions section, which is below too.
I could suggest the following as alternative (other proposals are welcome of course):

> `The primary use case is to allow Tango Device Server programmers to create Attributes and/or Commands in runtime.`

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: which => with
or
create generic Tango device classes which interfaces **are** entirely/partly unknown at the moment of creation
?

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: Good example!

I would suggest to add other use cases:

Consider a Tango Device Class that represents a residual gaz analyzer. The device will provide measurement results for many different masses. The results can be presented as a Tango spectrum attribute showing the measurements for all the masses but it can be also very interesting to create an attribute for each different mass, then the user could monitor this specific mass or define some alarm thresholds or configure archiving events for this specific mass. Dynamic attributes may be used to create easily these numerous attributes whose behaviour is almost identical.

Consider a Tango Device my/simu/device simulating the interface of another Tango Device my/original/device, for tests purposes. my/simu/device could query the interface of my/original/device at runtime and create dynamically the same attributes and commands as my/original/device. The simulated device could then be used to test high level client applications which would then talk to the simulated device instead of the original device.

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: What about attribute and command removal?

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]:
```suggestion
editor:  Reynald Bourtembourg `<at esrf.fr - reynald.bourtembourg>`
```

**Review by bourtemb — COMMENTED** [review]
Thank you @HelgaMond and @ingvord for the good work done on this RFC.
Please have a look at my comments.

File: 15/README.md
Comment (bourtemb) [reply 15/README.md]: I think something like: 

> Tango kernel library MUST provide API to server client to remove dynamic Attributes, Commands.

is missing.

Do we create another section dedicated to the removal?

> ## API that allows to remove Attributes and Commands in runtime

Or do we change the title:

> ## API that allows to create Attributes and Commands in runtime

into 
> ## API that allows to create and remove Attributes and Commands in runtime

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: I would be less radical and propose to replace MUST with MAY.

File: 15/README.md
Comment (bourtemb) [inline 15/README.md]: Just a general question to the community and RFC experts:
Do we consider that a valid Tango implementation is allowed to NOT implement the INTERFACE_CHANGE events as the MAY verb is suggesting here?

**Comment by bourtemb** [comment]
There is a conflict on the README.md file.
Could you please solve it?

File: 15/README.md
Comment (HelgaMond) [reply 15/README.md]: Any "static" can be defined via "dynamic" in implementation. Therefore there is a "MUST" word to enforce to use "dynamic".

**Comment by HelgaMond** [comment]
@bourtemb Thank you for the comments! The conflicts have been resolved.

**Comment by bourtemb** [comment]
> @bourtemb Thank you for the comments! The conflicts have been resolved.

It seems like some changes in the root README.md have been lost during this conflicts resolution.
For instance, the name of the README files in the RFC directories have been renamed from 2/README.md to 2/Device.md, etc...

According to that, I think 15/README.md file in the pull request should also be renamed.

**Review by bourtemb — APPROVED** [review]
Thanks @HelgaMond 
I still don't like much the note on V10 which is imposing a way to implement the features but since this is V9 specifications and not V10 specifications, I guess this is not that relevant here and I can live with that.

So, it looks good to me.
Maybe we should update the copyright year? But I guess that's not that important.
If someone else wants to review, I will wait a bit before merging.

**Review by bourtemb — DISMISSED** [review]
Ooops.
The last changes I asked were not yet implemented.
I'll wait for these changes...
> It seems like some changes in the root README.md have been lost during this conflicts resolution.
> For instance, the name of the README files in the RFC directories have been renamed from 2/README.md to 2/Device.md, etc...
> 
> According to that, I think 15/README.md file in the pull request should also be renamed.

**Comment by HelgaMond** [comment]
Dear @bourtemb Thank you for the comments! Could you please check if the problems with the resolution are solved? Thank you!

**Review by bourtemb — COMMENTED** [review]
Thanks @HelgaMond !
Looks good to me.
I guess it would be convenient to add links to the other related RFCs too.
Let's wait for @piogor 's answer on the convention for the links.
I also noticed some inconsistencies when using Upper case letters as first letter of some words (Tango Device Class vs Tango device class for instance)

**Comment by piogor** [comment]
> Let's wait for @piogor 's answer on the convention for the links.

Hi @bourtemb and @HelgaMond, I didn't realise the question, however, the convention is like the following: `[RFC 2/Device](/2)`
Is it what you are asking for?

**Comment by bourtemb** [comment]
> Hi @bourtemb and @HelgaMond, I didn't realise the question, however, the convention is like the following: `[RFC 2/Device](/2)`
> Is it what you are asking for?

Yes, thanks for the answer @piogor . So, it is not like in the main Readme.md file where we link to the md file directly?
[RFC-2](2/Device.md)

**Comment by piogor** [comment]
> Yes, thanks for the answer @piogor . So, it is not like in the main Readme.md file where we link to the md file directly?
> [RFC-2](2/Device.md)

We may need to synchronise the link convention in all the RFCs, later. However, for the moment, let's stay like this.

**Comment by piogor** [comment]
Added this: #88

**Review by bourtemb — COMMENTED** [review]
Thanks @HelgaMond  
As for the Pipe RFC, the content looks good to me.
There is still an outstanding work to be done to add links to other RFCs but this can be done later on when a decision will have been taken in https://github.com/tango-controls/rfc/issues/88.
The question of when using upper case first letters can also be addressed later on when a global harmonization work on all the RFCs will be done.
So, if no objection, I will merge this PR this afternoon after the RFC meeting.

**Comment by HelgaMond** [comment]
@bourtemb Thank you very much! Sure, please, merge the RFC.

**Pull request merged**
